### PR TITLE
Tasks dev

### PR DESCRIPTION
**Pull request by noacoh**

### CODE REVIEW

File: models/task.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+const taskSchema = new Schema({
+    id: Number,
+    courseId: Number,
+    title: String,
+    filePath: String,
+    solutionPath: String,
+    created: Date,
+    deadline: Date
+});
```
Comment (noacoh) [inline models/task.js]: add course field as ref

File: routes/tasks.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+const TasksController = require('../controllers/tasks');
+const router = require('express-promise-router')();
+
+router.route('/')
+    .get(TasksController.index)
+    .post(TasksController.newTask);
```
Comment (noacoh) [inline routes/tasks.js]: add body validation with joi
don't forget to extract body from req.value.body in the controller function

File: controllers/tasks.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const Task = require('../models/task');
+
+module.exports = {
+  index: async (req, res, next) => {
+    try{
```
Comment (noacoh) [inline controllers/tasks.js]: remove try catch block

File: controllers/tasks.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        const tasks = await Task.find({});
+        res.status(200).json(tasks);
+    } catch(err){
+          next(err);}
+  },
+    newTask: async (req, res, next) => {
+      try{
```
Comment (noacoh) [inline controllers/tasks.js]: remove try catch block

File: controllers/submissions.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    index: async (req, res, next) => {
+        const submissions = await Submission.find({});
+        res.status(200).json(submissions);
+    },
+    getSubmission: async (req, res, next) => {
+        const { submissionId } = req.value.params;
+        const submission = await Submission.findById(submissionId);
+        res.status(200).json(submission);
+    },
+    getTaskSubmissions: async (req, res, next) => {
```
Comment (noacoh) [inline controllers/submissions.js]: This should be on controllers/tasks

const { taskId } = req.value.params;
const task = await Task.findById(taskId).populate('submissions');
res.status(200).json(task.submissions);

File: controllers/submissions.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        const { taskId } = req.value.params;
+        const submissions = await Submission.findById(taskId).toArray();
+        res.status(200).json(submissions);
+    },
+    getUserSubmissions: async (req, res, next) => {
```
Comment (noacoh) [inline controllers/submissions.js]: Remove this

File: models/submission.js [hunk 7]
```diff
@@ -0,0 +1,19 @@
+const mongoose = require('mongoose');
+const Schema = mongoose.Schema;
+
+const
+    submissionSchema = new Schema({
+    submissionDate: Date,
+    grade: Number,
+    filePath: String,
+    task: [{
```
Comment (noacoh) [inline models/submission.js]: this should be a ref to only one task

File: models/submission.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+        type: Schema.Types.ObjectId,
+        ref: 'task'
+    }],
+    student: [{
```
Comment (noacoh) [inline models/submission.js]: this should ref only to one student (notice student not user )

File: models/task.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+const Schema = mongoose.Schema;
+
+const taskSchema = new Schema({
+    id: Number,
+    title: String,
+    filePath: String,
+    solutionPath: String,
+    created: Date,
+    deadline: Date,
+    course: [{
```
Comment (noacoh) [inline models/task.js]: this should ref only one course

File: models/task.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+        type: Schema.Types.ObjectId,
+        ref: 'course'
+    }],
+    submissions: [{
```
Comment (noacoh) [inline models/task.js]: name this field 'studentSubmissions'

File: routes/submissions.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+router.route('/')
+    .get(SubmissionController.index);
+
+router.route('/:submissionId')
+    .get([validateParam(schemas.idSchema, 'submissionId'),
+        validateBody(schemas.submissionSchema)],
+        SubmissionController.getSubmission);
+
+router.route('/:taskId/submissions')
```
Comment (noacoh) [inline routes/submissions.js]: remove

File: routes/submissions.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    .get(validateParam(schemas.idSchema, 'taskId'),
+        SubmissionController.getTaskSubmissions);
+
+router.route('/:userId/submissions')
```
Comment (noacoh) [inline routes/submissions.js]: remove

File: helpers/routeHelpers.js [hunk 13]
```diff
@@ -39,26 +39,47 @@ module.exports = {
...
+            filePath: joi.string().required(),
+            solutionPath: joi.string().required(),
+            created: joi.date().required(),
+            deadline: joi.date().required()
+        }),
+        submissionSchema: joi.object().keys({
+            submissionDate: joi.string().required(),
+            grade: joi.number().required(),
+            filePath: joi.string().required(),
+
```
Comment (noacoh) [inline helpers/routeHelpers.js]: remove redundant line

File: helpers/routeHelpers.js [hunk 14]
```diff
@@ -39,26 +39,47 @@ module.exports = {
...
+            title: joi.string().required(),
+            filePath: joi.string().required(),
+            solutionPath: joi.string().required(),
+            created: joi.date().required(),
+            deadline: joi.date().required()
+        }),
+        submissionSchema: joi.object().keys({
+            submissionDate: joi.string().required(),
+            grade: joi.number().required(),
+            filePath: joi.string().required(),
```
Comment (noacoh) [inline helpers/routeHelpers.js]: missing fields-
student 
task

File: package.json [hunk 15]
```diff
@@ -12,6 +12,7 @@
     "express": "^4.17.0",
     "express-promise-router": "^3.0.3",
     "helmet": "^3.18.0",
+    "joi": "^14.3.1",
```
Comment (noacoh) [inline package.json]: remove joy thanks

File: controllers/submissions.js (hunk 5 above)
Comment (liranZiv27) [reply controllers/submissions.js]: done

**Pull request merged**